main.go

```diff
@@ -119,47 +119,110 @@ func newMemo() *memo {
 }
 
 func (m *memo) build(n *node) {
-	switch n.typ {
-	case join:
+	switch n.op.typ() {
+	case joinNode:
 		m.build(n.left)
 		m.build(n.right)
 		m.add(n)
 
-	case scan:
+	case scanNode:
 		m.add(n)
 	}
 }
 
+type xform interface {
+	check(n *node) bool
+	apply(n *node) []*node
+}
+
 // A ⋈ B => B ⋈ A
-func (m *memo) commute(n *node) *node {
-	if n.typ != join {
+type joinCommuteXform struct{}
+
+func (jc *joinCommuteXform) check(n *node) bool {
+	if n.op.typ() != joinNode {
+		return false
+	}
+	return true
+}
+
+func (jc *joinCommuteXform) apply(n *node) []*node {
+	if !jc.check(n) {
 		return nil
 	}
-	return &node{
-		typ:   join,
+	return []*node{{
+		op:    &join,
 		class: n.class,
 		left:  n.right,
 		right: n.left,
-	}
+	}}
 }
 
+var joinCommute joinCommuteXform
+
 // (A ⋈ B) ⋈ C  => A ⋈ (B ⋈ C)
-func (m *memo) associate(n *node) *node {
-	if n.typ != join || n.left.typ != join {
+type joinAssocXform struct{}
+
+func (ja *joinAssocXform) check(n *node) bool {
+	if n.op.typ() != joinNode || n.left.op.typ() != joinNode {
+		return false
+	}
+	return true
+}
+
+func (ja *joinAssocXform) apply(n *node) []*node {
+	if !ja.check(n) {
 		return nil
 	}
-	return &node{
-		typ:   join,
-		class: n.class,
-		left:  n.left.left,
-		right: &node{
-			typ:   join,
-			left:  n.left.right,
-			right: n.right,
+	r := &node{
+		op:    &join,
+		left:  n.left.right,
+		right: n.right,
+	}
+	return []*node{
+		r,
+		{
+			op:    &join,
+			class: n.class,
+			left:  n.left.left,
+			right: r,
 		},
 	}
 }
 
+var joinAssoc joinAssocXform
+
+type operator interface {
+	typ() nodeType
+	compatXform() []xform
+}
+
+type joinOp struct{}
+
+func (jo *joinOp) typ() nodeType {
```

Comment:
FYI, you can define these methods on `joinOp`. And since they don't actually use the receiver, the methods can look like: `func (joinOp) typ() nodeType ...`.